Add mechanism for sharing APIs between separately built hex files

oyvindronningstad opened the pull request:
This utilizes the fw_metadata from the bootloader subsys to share versioned APIs/ABIs between images.

An image puts two pointers into the fw_metadata:
 - The "out" API, which points to a function for retrieving its API.
 - The "in" API, which points to a noinit RAM section into which whoever is booting the image can copy a pointer to a similar retrieval function.

If all retrieval functions support it, APIs can be relayed between images that are more distant from each other in the boot chain/tree.

This mechanism is then used to share cryptographic functions from the first stage to the second stage, which will allow MCUboot to be slimmed down.

This depends on PR https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/109, but can be made to be independent via a temphack in nrf-connect-zephyr if desired.

oyvindronningstad commented:
cc @sigvartmh @hakonfam @osaether

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## Gitlint issues

Commit 7d7d4c7551:
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Update manifest"

Commit 32c95f02b6:
4: UC4 Line exceeds max length (73>72): " - Build bl_crypto test with bootloader in front. Link bl_crypto with app"

oyvindronningstad commented:
cc @frkv

hakonfam commented on subsys/bootloader/include/fw_metadata.h:
```diff
@@ -24,9 +24,17 @@
 
 #define MAGIC_LEN_WORDS (CONFIG_SB_MAGIC_LEN / sizeof(u32_t))
 
+#define TYPE_AND_DECL(retval, name, ...) \
```
An explanation as to what this does an why it is needed would be great.

oyvindronningstad replied:
Will add

oyvindronningstad replied:
Changed the name to `EXTABI_FUNCTION` and added docs.

hakonfam commented on subsys/bootloader/include/fw_metadata.h:
```diff
@@ -24,9 +24,17 @@
+struct fw_abi_getter_info;
+
 struct __packed fw_firmware_info {
 	u32_t magic[MAGIC_LEN_WORDS];
```
Once we are here, should this be documented?

oyvindronningstad replied:
Document the magic word? I could add something like

    /* Magic word to guarantee that this is an instance of fw_firmware_info. */

hakonfam replied:
I just wanted to have consistency since the other fields are documented. Perhaps the variable name in itself is sufficient, its up to you.

oyvindronningstad replied:
Added docs

sigvartmh commented on subsys/bootloader/fprotect/fprotect_bprot.c:
```diff
@@ -38,11 +38,7 @@ int fprotect_area(u32_t start, size_t length)
-	for (u32_t i = 0; i < BPROT_CONFIGS_NUM; i++) {
-		config_masks[i] = 0;
-	}
+	u32_t config_masks[BPROT_CONFIGS_NUM] = {0};
```
Maybe not include this in the PR

oyvindronningstad replied:
Will remove

sigvartmh commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+	return get_bl_crypto_abi()->abi.bl_sha256_finish(ctx, output);
+}
+
+int bl_ecdsa_verify_secp256r1(const u8_t * hash, u32_t hash_len, const u8_t * public_key, const u8_t * signature)
```
This is mostly my fault for not following the file's convention. Anyways I think `pk` is a bad abbreviation as it could be `public` or `private` key so I would suggest having `pubk` and `privk`  or write out key `pubkey` if we are going to abbreviate.

oyvindronningstad replied:
Done

sigvartmh commented on subsys/bootloader/bl_crypto/bl_crypto_oberon_ecdsa.c:
```diff
@@ -10,21 +10,46 @@
+	 */
+	if(hash_len != 32)
+	{
+		return 0xf00dbabe;
```
I added this for debugging I don't think it's strictly needed? Maybe should be an assert instead.

oyvindronningstad replied:
removed

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_cc310_ecdsa.c:
```diff
@@ -10,34 +10,47 @@
+}
+
+
+bool _verify_sig(const u8_t *data, u32_t data_len, const u8_t *sig,
```
Will add comments

oyvindronningstad replied:
I changed this around a bit.

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
Will add

oyvindronningstad replied:
Done

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+}
+
+int crypto_root_of_trust(const u8_t *pk, const u8_t *pk_hash,
+			 const u8_t *sig, const u8_t *fw,
```
> Or is this function so tightly coupled with parsing firmware data that it is more clear to use `fw`?

Basically, yes, that is the intention.

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+
+int crypto_root_of_trust(const u8_t *pk, const u8_t *pk_hash,
+			 const u8_t *sig, const u8_t *fw,
+			 const u32_t fw_len)
```
Well... This is such a critical function that I wanted to take the precaution.

hakonfam replied:
So we agree that there is no functional reason to have it like this.
Will the compiler start complaining if I throw a variable in here which is not `const` declared? If so, that will be a bit annoying I guess.

oyvindronningstad replied:
It prevents someone from changing the implementation to assign a new value to `fw_len`. It shouldn't require inputs to be const.

hakonfam replied:
But the callee can never changed the value passed by the caller (as its not a pointer), so this just limits what the implementation can use this variable for. 

For me its fine either way, I get your point.

sigvartmh commented on tests/subsys/bootloader/bl_crypto/fw_data.py:
```diff
@@ -0,0 +1,3114 @@
+fw_data = [
```
yes, actually what would have been best if this bin file is generated and used every-time as this is just a dump of the firmware which was supposed to be booted by MCUBoot to have a real-life test vector.

oyvindronningstad replied:
I tried to merge `test_generator.py` and `create_bl_crypto_test_vector.py` as well as `test_vector.c` and `test_vectors.h`, but I couldn't quite get it to work. Can you take a look @sigvartmh?

sigvartmh commented on subsys/bootloader/bl_crypto/bl_crypto_cc310_common.c:
```diff
@@ -14,34 +15,6 @@
-	return buf;
-}
-
-void *__weak memcpy(void *restrict d, const void *restrict s, size_t n)
```
not needed anymore? Is this because we use stdlib from zephyr?

hakonfam replied:
I think it was moved to a different file. Think it was mentioned in one of the commit messages

sigvartmh replied:
Ah yes found them in `bootloader.c`

oyvindronningstad commented on subsys/bootloader/fw_metadata/fw_metadata.c:
```diff
@@ -5,15 +5,68 @@
 
 extern const u32_t _image_rom_start;
 extern const u32_t _flash_used;
+extern const struct fw_firmware_info _firmware_info_start;
```
Good point, will do

oyvindronningstad replied:
Done

oyvindronningstad commented on subsys/bootloader/fw_metadata/fw_metadata.c:
```diff
@@ -5,15 +5,68 @@
+extern const struct fw_abi_info _extabis_start;
+extern const struct fw_abi_info _extabis_end;
+
+#define EXTABIS_INC(extabi) ((const struct fw_abi_info *)((u32_t)extabi + extabi->abi_len))
```
Yep, will add

oyvindronningstad replied:
Done

oyvindronningstad commented on subsys/bootloader/fw_metadata/fw_metadata.c:
```diff
@@ -5,15 +5,68 @@
+
+int abi_getter(u32_t id, u32_t index, const struct fw_abi_info ** abi)
+{
+	if (!abi) return -EFAULT;
```
One line `if`s not allowed?

hakonfam replied:
Zephyr follows Linux, and Linux states:
```
Don’t put multiple statements on a single line unless you have something to hide:
if (condition) do_this;
```

Zephyr coding style states:
```
Add braces to every if and else body, even for single-line code blocks.
```

oyvindronningstad replied:
Thanks, will change

oyvindronningstad replied:
Done

oyvindronningstad commented on subsys/bootloader/fw_metadata/fw_metadata.c:
```diff
@@ -5,15 +5,68 @@
+
+void abi_provide(u32_t address)
+{
+	printk("abi_provide");
```
No, will remove.

oyvindronningstad commented on subsys/bootloader/include/bl_crypto.h:
```diff
@@ -8,13 +8,38 @@
+typedef nrf_cc310_bl_hash_context_sha256_t bl_sha256_ctx_t;
+#else
+#define SHA256_CTX_SIZE 128
+typedef u32_t bl_sha256_ctx_t[SHA256_CTX_SIZE/4]; // Must be aligned.
```
It's a bit tricky since this is only a typedef. The comment is as much an explanation as an instruction. I will change the wording a bit.

oyvindronningstad replied:
Done

oyvindronningstad commented on subsys/bootloader/include/linker.ld:
```diff
@@ -13,7 +13,6 @@
 
 #define _LINKER
 #define _ASMLANGUAGE
-
```
Will remove

oyvindronningstad replied:
Done

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/src/main.c:
```diff
@@ -8,16 +8,275 @@
+void test_ecdsa_verify(void)
+{	
+	u32_t hash_len = ARRAY_SIZE(hash_sha256);	
+	/* All is good */
```
Will change to

    /* Success */

:)

oyvindronningstad replied:
Done

frkv commented on subsys/bootloader/bl_crypto/bl_crypto_cc310_common.c:
```diff
@@ -54,18 +27,18 @@ void cc310_bl_backend_disable(void)
 	NRF_CRYPTOCELL->ENABLE = 0;
```
I would assume this to be 1 when enabled

frkv commented on subsys/bootloader/bl_crypto/bl_crypto_internal.h:
```diff
@@ -41,8 +31,8 @@ bool verify_hash(const u8_t *data, size_t data_len, const u8_t *expected);
  */
-bool verify_truncated_hash(const u8_t *data, size_t data_len,
-			   const u8_t *expected, size_t hash_len);
+bool verify_truncated_hash(const u8_t *data, u32_t data_len,
```
bool external not documented (in all definitions in this file)

oyvindronningstad replied:
Will add docs

oyvindronningstad replied:
I amended the docs in this file.

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/test_generator.py:
```diff
@@ -0,0 +1,157 @@
+    return ret_str[:-1]
+
+fw_data_file = open('fw_data.py')
+exec(fw_data_file.read())
```
I migrated fw_data to a bin file, so this has been removed

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/test_generator.py:
```diff
@@ -0,0 +1,157 @@
+fw_sig = b'ac95651230dee1b857d29971fd5177cf4536ee4a819abaec950cccae27548a3823ff093cc2a64a8dab7f4df73dec98'
+
+assert vk.verify(signature, my_hash)
+test_vectors.code.append(C.variable("pub_x", typename='u8_t', array=32))
```
I migrated it to the method used in `create_bl_crypto_test_vector.py`, so this is gone.

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+#include <kernel.h>
+
+extern struct fw_abi_getter_info abi_getter_in;
+void print_header(const struct bl_crypto_abi * obj){
```
Removed the function

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+	printk("ABI len: %d\n\r", obj->header.abi_len);
+}
+
+const struct bl_crypto_abi* get_bl_crypto_abi(void)
```
Should be consistent across the patch now.

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+		int ret = abi_getter_in.abi_getter(0x12345678, 0, &bl_crypto_abi);
+		if (ret) k_oops();
+	}
+	//print_header((const struct bl_crypto_abi *)bl_crypto_abi);
```
Done

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_client.c:
```diff
@@ -0,0 +1,69 @@
+
+const struct bl_crypto_abi* get_bl_crypto_abi(void)
+{
+	//printk("Getting crypto ABI.\n");
```
Done

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_oberon_ecdsa.c:
```diff
@@ -10,21 +10,46 @@
 	return (retval == 0);
 }
+
+/* Returns 0 for succes or -1 for failure */
```
Done

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_oberon_ecdsa.c:
```diff
@@ -10,21 +10,46 @@
+							  const u8_t * signature)
+{
+	int retval;
+	/* maybe we need internal data
```
Done

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_oberon_hash.c:
```diff
@@ -19,23 +24,35 @@ bool get_hash(u8_t *hash, const u8_t *data, u32_t data_len)
+{
+	occ_sha256_final(output, ctx);
+	return 0;
+}
```
Done

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/prj.conf:
```diff
@@ -10,3 +10,4 @@ CONFIG_TEST_USERSPACE=n
 CONFIG_USERSPACE=n
 CONFIG_STDOUT_CONSOLE=n
 CONFIG_SECURE_BOOT=y
+#CONFIG_MULTI_IMAGE_SHARED_SRAM=y
```
Done

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/src/main.c:
```diff
@@ -8,16 +8,275 @@
 #include "test_vector.c"
+#include "test_vectors.h"
+
+/*
```
Done

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/src/main.c:
```diff
@@ -8,16 +8,275 @@
+	zassert_equal(0, retval, "const_pk const_fw retval: 0x%x", retval);
+	
+	
+	/*
```
Done

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/src/main.c:
```diff
@@ -8,16 +8,275 @@
+	for(size_t i = 0; i < 32; i++) {
+		printk("%x",test_vector[i]);
+	}
+	printk("\n\ro vector: 0x");
```
Done

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/test_generator.py:
```diff
@@ -0,0 +1,157 @@
+#secp256r1 = prime256v1 = NIST P-256
```
Removed

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/test_generator.py:
```diff
@@ -0,0 +1,157 @@
+long_input_hash = hexstr_to_array(sha256(long_input).hexdigest())
+long_input = hexstr_to_array(long_input.hex())
+
+#fw_hash = b'f1ab54f3f73164d9990159bbd341ec74789caaff14fff7d26ddab56da44a3'
```
Removed

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/test_vector.c:
```diff
@@ -3,30 +3,59 @@
 	0xa1, 0x8d, 0xbf, 0x87, 0x7e, 0x68};
 
-uint8_t firmware_hash[] = {0x53, 0x2d, 0x74, 0xb1, 0x5, 0xaf, 0x9d, 0xc6, 0x0,
+//static uint8_t firmware_hash[] = {0x53, 0x2d, 0x74, 0xb1, 0x5, 0xaf, 0x9d, 0xc6, 0x0,
```
Done

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/test_generator.py:
```diff
@@ -0,0 +1,157 @@
+print(fw_sig_hex.hex())
+gen_sig = hexstr_to_array(generated_sig.hex())
+
+fw_x = fw_vk.pubkey.point.x()
```
I moved the code into an `if __name__ == "__main__":`. Is that what you meant?

oyvindronningstad commented on tests/subsys/bootloader/bl_crypto/test_generator.py:
```diff
@@ -0,0 +1,157 @@
+fw_sig = [0xac, 0x95, 0x65, 0x12, 0x30, 0xde, 0xe1, 0xb, 0x8, 0x57, 0xd2, 0x9, 0x97, 0x1f, 0xd5, 0x17, 0x7c, 0xf4, 0x53, 0x6e, 0xe4, 0xa8, 0x19, 0xab, 0xae, 0xc9, 0x50, 0xcc, 0xca, 0xe2, 0x75, 0x48, 0xa3, 
+0x82, 0x3f, 0xf0, 0x93, 0xcc, 0x2a, 0x64, 0xa8, 0xd, 0xab, 0x7f, 0x4d, 0xf7, 0x3d, 0xec, 0x9a, 0xac, 0x4, 0x72, 0x54, 0x2d, 0x55, 0xee, 0xca, 0xc0, 0x79, 0xad, 0x2c, 0x6a, 0xee, 0x58]
+fw_sig_hex = arr_to_hexstr(fw_sig)
+print(sha256(fw_hex).hexdigest())
```
Removed `print`s

oyvindronningstad commented:
@SebastianBoe @ioannisg Would be great to get your feedback on the sharing mechanism.

SebastianBoe commented on subsys/bootloader/fw_metadata/fw_metadata.c:
```diff
@@ -5,15 +5,68 @@
+
+void abi_provide(u32_t address)
+{
+	const struct fw_firmware_info *our_fw_info = _firmware_info_start;
```
is _firmware_info_start the same as &m_firmware_info ?

SebastianBoe replied:
If so, we could drop the custom section, or?

oyvindronningstad replied:
No. There are actually two custom sections at play: `.firmware_info` has been around for a while and needs to be placed at a specific offset inside the image (currently 0x800). `extabis` has been added for this patch and is meant as an aggregator for exposed abis. Any module can declare an ABI inside the extabis sections, and the list will be provided to the next image.

SebastianBoe replied:
Ah, my mistake, I assumed Z_GENERIC_SECTION was declaring a section somehow, and also adding the .firmware_info attribute to m_firmware_info.

oyvindronningstad replied:
That would've been nice :)

SebastianBoe commented on subsys/bootloader/include/fw_metadata.h:
```diff
@@ -106,9 +127,9 @@ OFFSET_CHECK(struct fw_validation_pointer, validation_info,
-	      && !((u32_t)len & 3),
+	__ASSERT(!((u32_t)expected % 4)
+	      && !((u32_t)actual % 4)
+	      && !((u32_t)len % 4),
```
don't show this to @osaether  ;)

hakonfam replied:
Haha

SebastianBoe commented on subsys/bootloader/include/fw_metadata.h:
```diff
@@ -54,6 +70,11 @@ OFFSET_CHECK(struct fw_firmware_info, firmware_version,
 
+
+#define ABI_GETTER_OUT_NAME abi_getter_out
+#define ABI_GETTER_IN_NAME  abi_getter_in
```
I'd appreciate a comment explaining this.

At first glance the indirection seems unnecessary.

Either we are doing something fancy, or we want to be able to just these names in the future for some reason?

oyvindronningstad replied:
This has been removed

SebastianBoe commented on subsys/bootloader/bl_crypto/bl_crypto_oberon_hash.c:
```diff
@@ -9,8 +9,14 @@
 
-bool get_hash(u8_t *hash, const u8_t *data, u32_t data_len)
+
+static_assert(SHA256_CTX_SIZE >= sizeof(occ_sha256_ctx), \
```
BUILD_ASSERT

oyvindronningstad replied:
Done

SebastianBoe commented:
I'm sorry, I don't have the capacity to review this any time soon.

But the build and configuration changes look fairly straightforward.

pizi-nordic commented:
How are you going to manage RAM of the API exporter? Is the API stateless?

frkv commented:
Readability point:

I'm wondering if extabi is unnecessarily complex grouping of characters for an invented concept. Is there a possibility to add an underscore in the naming to give a natural break when parsing it?

This would mean renaming keywords and macros as such: __ext_abi, EXT_ABI_XXXX etc.

oyvindronningstad commented:
I just pushed quite a few changes, so here's a summary:

- crypto_root_of_trust -> root_of_trust_verify
- extabi -> ext_abi
- static_assert -> BUILD_ASSERT_MSG
- Cleaned up magic number to more closely resemble actual coupling: i.e. decouple fw_info and validation_info, and couple fw_info and abi_info.
- Removed abi_getter_info struct. fw_info now links directly to the getter function.
- Entries in the .ext_abis section are now pointers to abi_info structs, so iteration is easier and less prone to alignment issues.
- Divided the ABI into 3: "rot_verify", "sha256", and "secp256r1".
- Increased stack usage by the ABIs to 4k to improve speed.
- Depend on an nrf-connect-zephyr temphack instead of on the full linker snippets PR.
- Other small improvements and fixes

oyvindronningstad commented:
> How are you going to manage RAM of the API exporter? Is the API stateless?

Yes, the assumption is that the API uses only stack memory and flash. Any other state must be handled via some shared RAM mechanism, but this is out of scope for this PR.

sigvartmh commented on subsys/bootloader/bl_crypto/bl_crypto_cc310_hash.c:
```diff
@@ -7,94 +7,200 @@
 }
 
-bool get_hash(u8_t *hash, const u8_t *data, u32_t data_len)
+int bl_sha256_abort(nrf_cc310_bl_hash_context_sha256_t *ctx)
```
What is the final goal of this function?

oyvindronningstad replied:
It's part of the ARM PSA crypto API.

sigvartmh replied:
Okei, nice job keeping it future compatible :)

oyvindronningstad replied:
Come to think of it, with versioning of ABIs, it might be better to leave it out as long as it's not implemented, and rather increment the version when it's actually implemented.

oyvindronningstad replied:
What do you say @frkv ?

frkv replied:
Abort is only practical in a multi-threaded operation and/or in a scenario where this is accessed from an NSPE. I don't know how far this "shared interface" is going to be extended. If this is exposed in a non-secure domain then the abort makes sense, but until that point the API is probably a bit superfluous

oyvindronningstad replied:
I removed `abort` now. I think not including it will make things less confusing.

oyvindronningstad commented:
> Readability point:
> 
> I'm wondering if extabi is unnecessarily complex grouping of characters for an invented concept. Is there a possibility to add an underscore in the naming to give a natural break when parsing it?
> 
> This would mean renaming keywords and macros as such: __ext_abi, EXT_ABI_XXXX etc.

Done

hakonfam commented:
Sorry for being late, but I guess @frkv  already has done his review.

If this looks OK for you @frkv  I will do the merge.

Have you tested this with latest ncs/master @oyvindronningstad  ?
If something is missing, could you list the PRs needed for this to work?

oyvindronningstad commented:
> Have you tested this with latest ncs/master @oyvindronningstad ?
This is rebased today. But give me the rest of the day to test a few more configurations.

> If something is missing, could you list the PRs needed for this to work?
It's probably best if this waits until https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/115 is merged or resolved in some other way.

sigvartmh commented on subsys/bootloader/bl_crypto/bl_crypto_cc310_hash.c:
```diff
@@ -10,91 +10,134 @@
+int bl_sha256_finish(nrf_cc310_bl_hash_context_sha256_t * ctx, u8_t * output)
+{
+	int retval = nrf_cc310_bl_hash_sha256_finalize(ctx,(nrf_cc310_bl_hash_digest_sha256_t *) output);
+	cc310_bl_backend_disable();
```
should the hash functions be wrapped in enable/disable? Isn't good enough to assume you would do init->update->Finalize or init->finalize?

frkv replied:
For the smallest possible power drain the CC310 HW should be enabled/disabled around the calls you are doing. not kept on between functions

sigvartmh replied:
That makes sense tnx for the reply 👍

oyvindronningstad commented on subsys/bootloader/include/fw_metadata.h:
```diff
@@ -176,4 +201,86 @@ validation_info_find(const struct fw_firmware_info *finfo, u32_t
+		"extabi " #type " has wrong offset for header." #member)
+
+#define __extabi(type) \
+	OFFSET_CHECK_EXTABI(type, magic, 0); \
```
Yes, but how could I do that?
I would like to have done something like

    STATIC_ASSERT(offsetof(type, header) == 0);
    STATIC_ASSERT(typeof(type, header) == "struct fw_abi_info");

but I don't know of anything resembling this. So these checks are the closest I could think of.

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto_cc310_hash.c:
```diff
@@ -10,91 +10,134 @@
+
+int bl_sha256_init(nrf_cc310_bl_hash_context_sha256_t * const ctx)
+{
+	cc310_bl_backend_enable();
```
So `bl_sha256_init` doesn't need `_enable` and `_disable`?

frkv replied:
Hash init does not started any HW operation. It will just set the SHA-256 initial values in the context object, so no...

oyvindronningstad replied:
Ok great, what about finalize?

frkv replied:
Finalize uses HW

oyvindronningstad replied:
Should be good now. backend is enabled for `update` and `finalize`.

oyvindronningstad commented on subsys/bootloader/bl_crypto/bl_crypto.c:
```diff
@@ -7,20 +7,88 @@
+		.bl_sha256_init = bl_sha256_init,
+		.bl_sha256_update = bl_sha256_update,
+		.bl_sha256_finish = bl_sha256_finish,
+		.bl_sha256_ctx_size = SHA256_CTX_SIZE,
```
I added a `bl_sha256_verify` but it might not have been what you meant?

oyvindronningstad replied:
Modified `bl_sha256_verify` now.

frkv replied:
Yes. Looks good

oyvindronningstad commented:
The latest push includes a bit of refactoring to transition bl_crypto function return values from `bool` to `int`.

hakonfam commented:
@oyvindronningstad FYI  the bootloader as app Pr is now in. As you can probably see from the conflicts you are having :)

oyvindronningstad commented:
@pizi-nordic This is nearing completion, any last comments?

hakonfam commented:
I'm OOO today, so send me an email when the conflict is fixed :)

lemrey commented on CODEOWNERS:
```diff
@@ -50,6 +50,7 @@
 /subsys/bootloader/                       @hakonfam @ioannisg
 /subsys/enhanced_shockburst/              @Raane @lemrey
 /subsys/event_manager/                    @pdunaj
+/subsys/fw_metadata/                      @hakonfam
```
Please split this change in a separate commit.

oyvindronningstad replied:
Can I ask why? The directory was created in that commit.

lemrey replied:
Adding a code-owner is done in a separate commit, for example: df9ccc74c37d692127e59c246a84cea12b5c4c20
I suppose that `CODEOWNERS` is a special file and it should be clear in the log when it is changed.

oyvindronningstad replied:
I'd rather not, since builds take long, and I want to merge this, I'll change it if I have to change something else, ok?

BTW, I don't think it's that strict: https://github.com/zephyrproject-rtos/zephyr/commit/6b21e1b7a739e95bd7a19800f72ef587e72c1a87#diff-5daf978e60e03b3975b485d5f1b449dc

oyvindronningstad replied:
Done, but let's not run CI again

oyvindronningstad commented:
Ready to merge.

sigvartmh commented:
> Ready to merge.

@carlescufi @SebastianBoe

lemrey commented on include/bl_crypto.h:
```diff
@@ -0,0 +1,189 @@
+
+
+#if CONFIG_SB_CRYPTO_OBERON_SHA256
+	#include <occ_sha256.h>
```
Don't indent pre-compiler directives, if possible.

oyvindronningstad replied:
It's totally unreadable without, so pretty please?

oyvindronningstad replied:
Ok, so can we merge @lemrey ?

lemrey commented on include/fw_metadata.h:
```diff
@@ -0,0 +1,241 @@
+	BUILD_ASSERT_MSG(offsetof(type, header.member) == value, \
+		"ext_abi " #type " has wrong offset for header." #member)
+
+#define __ext_abi(type, name) \
```
Don't care much here, but macros should be uppercase and not `__` prefixed since those symbols are, in theory, reserved for linker symbols. They have started to enforce this upstream because of MISRA-C compliance.

oyvindronningstad replied:
It's named like this because it's meant to be used the same way as e.g. `__noinit`

lemrey replied:
Ok, thanks.

oyvindronningstad commented:
> There is a bit of misalignment between the header files and the API they implement.
> For example:
> 
> * `bl_crypto.h` implements mostly `bl_*` prefixed functions, with the exception of `crypto_init` and `root_of_trusty_verify` which are named differently. Can we fix that?
> * `fw_metadata.h` implements `abi_*` APIs. Why not call it `abi.h` for example?

I did some renaming in the public bl_crypto.h and fw_metadata.h functions.

lemrey commented:
> I did some renaming in the public bl_crypto.h and fw_metadata.h functions.

Good, but have you missed `cryto_init` (-> `bl_crypto_init`) ? :)

oyvindronningstad commented:
> > I did some renaming in the public bl_crypto.h and fw_metadata.h functions.
> 
> Good, but have you missed `cryto_init` (-> `bl_crypto_init`) ? :)

Oh right :P, fixed now

oyvindronningstad commented:
@hakonfam Ready to merge, unless you had more comments @lemrey?

lemrey commented on subsys/bootloader/bl_crypto/CMakeLists.txt:
```diff
@@ -29,3 +29,6 @@ elseif (CONFIG_SB_CRYPTO_CC310_SHA256)
   message(FATAL_ERROR "No hash implementation chosen for bootloader.")
 endif()
+
+zephyr_include_directories(.)
```
This shouldn't be needed anymore.

oyvindronningstad replied:
You're right, I'll remove it in a later PR.
